### ISSUE TITLE
OpenCL fails on a Ryzen 5700G APU on ROCm 5.6.0 - regression (command queue creation fails with `CL_OUT_OF_HOST_MEMORY`)

### ISSUE THREAD

**Issue by joanbm** [RadeonOpenCompute/ROCR-Runtime#163]
## Description

After upgrading my Arch Linux system with a Ryzen 5700G APU to ROCm 5.6.0, running any simple OpenCL program (see below for an example) fails.

In particular, during initialization, `clCreateCommandQueueWithProperties` will return with error code -6 (`CL_OUT_OF_HOST_MEMORY`) and fail to create the command queue.

If I downgrade ROCm to 5.5.1, OpenCL applications run fine again.

Another workaround I found is setting the environment variable `HSA_SCRATCH_SINGLE_LIMIT=0`. At least from my testing this resolves the issue and I can launch various simple OpenCL applications and kernels successfully.

## Analysis

The problem seems related to the [`GpuAgent::ReserveScratch`](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/e0fadddb3175cb95ce9e9af2ebd2a205045e854e/src/core/runtime/amd_gpu_agent.cpp#L485) function that was introduced [between ROCm 5.5.1 and 5.6.0](https://github.com/RadeonOpenCompute/ROCR-Runtime/compare/rocm-5.5.1...rocm-5.6.0).

In particular, it appears that [by default it will try to allocate ~140MB of memory](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/e0fadddb3175cb95ce9e9af2ebd2a205045e854e/src/core/util/flag.h#L67C7-L67C7).

I suppose my Ryzen 5700G can not provide this much memory and that's why it fails. That's why `HSA_SCRATCH_SINGLE_LIMIT=0` works around the problem as it disables the above allocation.

Even setting `HSA_SCRATCH_SINGLE_LIMIT` to `8000000` (~8MB) will still cause `GpuAgent::ReserveScratch` to fail. Values of `7000000` (~7MB) usually work.

### Steps to Reproduce

* Save the following OpenCL example:

   ```cpp
   #include <iostream>
   #include <vector>
   
   #define CL_HPP_ENABLE_EXCEPTIONS
   #define CL_HPP_TARGET_OPENCL_VERSION 200
   #include <CL/opencl.hpp>
   
   int main() {
   	try {
   		std::vector<cl::Platform> platforms;
   		cl::Platform::get(&platforms);
   		if (platforms.empty()) return EXIT_FAILURE;
   
   		std::vector<cl::Device> devices;
   		platforms[0].getDevices(CL_DEVICE_TYPE_GPU, &devices);
   		if (devices.empty()) return EXIT_FAILURE;
   
   		cl::Device device = devices[0];
   		std::cout << "Vendor: " << device.getInfo<CL_DEVICE_VENDOR>() << std::endl;
   		std::cout << "Name: " << device.getInfo<CL_DEVICE_NAME>() << std::endl;
   
   		cl::Context context(device);
   		cl::CommandQueue queue(context, device);
   		std::cout << "Success!" << std::endl;
   		return EXIT_SUCCESS;
   	} catch (const cl::Error &err) {
   		std::cerr
   			<< "OpenCL error: "
   			<< err.what() << "(" << err.err() << ")"
   			<< std::endl;
   		return 1;
   	}
   }


    ```

* Build it:

   ```console
   $ g++ hello.cpp -o hello -lOpenCL
   ```

* Run it on a machine with a Ryzen 5700G processor and ROCm 5.6.0:

   ```
   $ ./hello
   ```

## Expected Behavior:

The program should run correctly and print the following output:

```
Vendor: Advanced Micro Devices, Inc.
Name: gfx90c:xnack-
Success!
```

## Actual Behavior

The program fails with a `CL_OUT_OF_HOST_MEMORY` on `clCreateCommandQueueWithProperties`:

```
Vendor: Advanced Micro Devices, Inc.
Name: gfx90c:xnack-
OpenCL error: clCreateCommandQueueWithProperties(-6)
```

## Environment:

* OS: Arch Linux updated as of 2023-08-06, Linux kernel 6.4.8.
* Hardware: Ryzen 5700G APU, MSI B450M Mortar motherboard, 32GB of RAM.

`clinfo` output:

```
Number of platforms                               1
  Platform Name                                   AMD Accelerated Parallel Processing
  Platform Vendor                                 Advanced Micro Devices, Inc.
  Platform Version                                OpenCL 2.1 AMD-APP.dbg (3570.0)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd cl_amd_event_callback 
  Platform Extensions function suffix             AMD
  Platform Host timer resolution                  1ns

  Platform Name                                   AMD Accelerated Parallel Processing
Number of devices                                 1
  Device Name                                     gfx90c:xnack-
  Device Vendor                                   Advanced Micro Devices, Inc.
  Device Vendor ID                                0x1002
  Device Version                                  OpenCL 2.0 
  Driver Version                                  3570.0 (HSA1.1,LC)
  Device OpenCL C Version                         OpenCL C 2.0 
  Device Type                                     GPU
  Device Board Name (AMD)                         AMD Radeon Graphics
  Device PCI-e ID (AMD)                           0x1638
  Device Topology (AMD)                           PCI-E, 0000:30:00.0
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               8
  SIMD per compute unit (AMD)                     4
  SIMD width (AMD)                                16
  SIMD instruction width (AMD)                    1
  Max clock frequency                             2000MHz
  Graphics IP (AMD)                               9.0
  Device Partition                                (core)
    Max number of sub-devices                     8
    Supported partition types                     None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             1024x1024x1024
  Max work group size                             256
  Preferred work group size (AMD)                 256
  Max work group size (AMD)                       1024
  Preferred work group size multiple (kernel)     <getWGsizes:1504: create kernel : error -6>
  Wavefront width (AMD)                           64
  Preferred / native vector sizes                 
    char                                                 4 / 4       
    short                                                2 / 2       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 1 / 1        (cl_khr_fp16)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     No
    Infinity and NANs                             No
    Round to nearest                              No
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              33554432 (32MiB)
  Global free memory (AMD)                        6144 (6MiB) 6144 (6MiB)
  Global memory channels (AMD)                    4
  Global memory banks per channel (AMD)           4
  Global memory bank width (AMD)                  256 bytes
  Error Correction support                        No
  Max memory allocation                           28521264 (27.2MiB)
  Unified memory for Host and Device              No
  Shared Virtual Memory (SVM) capabilities        (core)
    Coarse-grained buffer sharing                 Yes
    Fine-grained buffer sharing                   Yes
    Fine-grained system sharing                   No
    Atomics                                       No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Preferred alignment for atomics                 
    SVM                                           0 bytes
    Global                                        0 bytes
    Local                                         0 bytes
  Max size for global variable                    28521264 (27.2MiB)
  Preferred total size of global vars             33554432 (32MiB)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        16384 (16KiB)
  Global Memory cache line size                   64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             5688
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 8192 images
    Base address alignment for 2D image buffers   256 bytes
    Pitch alignment for 2D image buffers          256 pixels
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             16384x16384x8192 pixels
    Max number of read image args                 128
    Max number of write image args                8
    Max number of read/write image args           64
  Max number of pipe args                         16
  Max active pipe reservations                    16
  Max pipe packet size                            28521264 (27.2MiB)
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Local memory size per CU (AMD)                  65536 (64KiB)
  Local memory banks (AMD)                        32
  Max number of constant args                     8
  Max constant buffer size                        28521264 (27.2MiB)
  Preferred constant buffer size (AMD)            16384 (16KiB)
  Max size of kernel argument                     1024
  Queue properties (on host)                      
    Out-of-order execution                        No
    Profiling                                     Yes
  Queue properties (on device)                    
    Out-of-order execution                        Yes
    Profiling                                     Yes
    Preferred size                                262144 (256KiB)
    Max size                                      8388608 (8MiB)
  Max queues on device                            1
  Max events on device                            1024
  Prefer user sync for interop                    Yes
  Number of P2P devices (AMD)                     0
  Profiling timer resolution                      1ns
  Profiling timer offset since Epoch (AMD)        0ns (Thu Jan  1 00:00:00 1970)
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    Thread trace supported (AMD)                  No
    Number of async queues (AMD)                  8
    Max real-time compute queues (AMD)            8
    Max real-time compute units (AMD)             8
  printf() buffer size                            4194304 (4MiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_gl_sharing cl_amd_device_attribute_query cl_amd_media_ops cl_amd_media_ops2 cl_khr_image2d_from_buffer cl_khr_subgroups cl_khr_depth_images cl_amd_copy_buffer_p2p cl_amd_assembly_program 

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              Success [AMD]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx90c:xnack-
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx90c:xnack-
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx90c:xnack-
	NOTE:	your OpenCL library only supports OpenCL 2.0,
		but some installed platforms support OpenCL 2.1.
		Programs using 2.1 features may crash
		or behave unexpectedly
```

**Comment by dayatsin-amd** [comment]
Hi @joanbm  Thanks for the update. We have a patch coming in ROCm-5.7 that should fix this:

-----------------------
commit 281d6a7ee212792858b12649845d45ade4a071f2
Author: Xiaomeng Hou <Xiaomeng.Hou@amd.com>
Date:   Tue Jun 6 22:01:31 2023 +0800

    Do not reserve scratch memory on asic with finite vram resource

    Change-Id: I0a2207cb01f464ed3e73331637cfa9bd62f03d97

diff --git a/opensrc/hsa-runtime/core/runtime/amd_gpu_agent.cpp b/opensrc/hsa-runtime/core/runtime/amd_gpu_agent.cpp
index 7ac0c084..f4ac5ca7 100644
--- a/opensrc/hsa-runtime/core/runtime/amd_gpu_agent.cpp
+++ b/opensrc/hsa-runtime/core/runtime/amd_gpu_agent.cpp
@@ -521,8 +521,11 @@ void GpuAgent::InitScratchPool() {
 void GpuAgent::ReserveScratch()
 {
   size_t reserved_sz = core::Runtime::runtime_singleton_->flag().scratch_single_limit();
+  size_t available;
+  HSAKMT_STATUS err = hsaKmtAvailableMemory(node_id(), &available);
+  assert(err == HSAKMT_STATUS_SUCCESS && "hsaKmtAvailableMemory failed");
   ScopedAcquire<KernelMutex> lock(&scratch_lock_);
-  if (!scratch_cache_.reserved_bytes() && reserved_sz) {
+  if (!scratch_cache_.reserved_bytes() && reserved_sz && available > 8 * reserved_sz) {
     HSAuint64 alt_va;
     void* reserved_base = scratch_pool_.alloc(reserved_sz);
     assert(reserved_base && "Could not allocate reserved memory");
-----------------------

**Comment by joanbm** [comment]
Thanks! I just built the runtime with the patch applied and I can confirm that it fixes the problem.

**Comment by joanbm** [comment]
The fix was included on ROCm 5.7.0 and I can confirm it works on my system.
